### ISSUE TITLE
Storage of actor reminders in one single db-field

### ISSUE THREAD

**Issue by javageek79** [dapr/dapr#2889]
## In what area(s)?

/area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

## Ask your question here
During our tests, we realized that reminders of all virtual actors are getting persisted in one single `data`-Field of the actor-type db entry. 
Doesn't this bear the risk of becoming a bottle neck during runtime?
If there are e.g. 10.000 actors, each having 4 reminders, there will be 40.000 entries in that particular `data`-fied.

**Comment by withinboredom** [comment]
> During our tests, we realized that reminders of all virtual actors are getting persisted in one single data-Field of the actor-type db entry.

I believe this choice of data structure is most likely due to the state API's not (yet) supporting some query mechanisms.

**Comment by javageek79** [comment]
@artursouza as discussed,
regarding registering of reminders, we would see the following intervals in the above described load scenario
```
>= 3sec no go
2 <= delay < 3 occasionally acceptable
1 <= delay < 2 acceptable
< 1sec ok
```

**Comment by javageek79** [comment]
Hi @artursouza, we just applied the feature on our QA stage, where we hav multiple minreplicas. The migration warning in the logs, I can see multiple times. But I don't find any of the complete warnings. In the statestore, only the metadata entry is created.

Is there something  we need to take care of additionally or does the soloution deals with multiple cuncurrent migrations?

**Comment by artursouza** [comment]
> Hi @artursouza, we just applied the feature on our QA stage, where we hav multiple minreplicas. The migration warning in the logs, I can see multiple times. But I don't find any of the complete warnings. In the statestore, only the metadata entry is created.
> 
> Is there something we need to take care of additionally or does the soloution deals with multiple cuncurrent migrations?
> 
> I also tried with a new collection in the same database, but sidecar log says
> 
> ```
> error getting reminders for actor type MyActor_v2: could not read actor type metadata: 
> BadRequest, Message: {
>   "Errors": [ "Encountered exception while executing function. Exception = Error: {
>     \"Errors\":[\"Requests originating from scripts cannot reference partition keys other than the one for which client request was submitted.\"]}\r\n
>       Stack trace: Error: {\"Errors\":[\"Requests originating from scripts cannot reference partition keys other than the one for which client request was submitted.\"]}\n   
>       at callback (script.js:51:18)\n   
>      at Anonymous function (script.js:760:29)"
> ]}
> ```


The code takes care of concurrent migrations. This issue has to do with how the code writes to the database since CosmosDB needs a partitionKey. I am fixing this now.

**Comment by artursouza** [comment]
Fix is in the PR above. It will be part of the 1.4 release next week. Since this is a preview feature, this fix is specific to one state store and the 1.4 release is just around the corner, we don't plan to release a patch for 1.3. Please, let me know if this plan works for you.

**Comment by javageek79** [comment]
@artursouza, yes, this should be fine, thx

**Comment by javageek79** [comment]
We still face some issues on our QA stage.

1. As partiotions can only be in creased, we started with a lower number of partiotions. Unfortunatelly, the number was too small, so that we ran into the same "Request size too large" issue in some of the partiotions. When increasing the number of partitions in such a state, the migration fails due to the same "Request size too large" error.

2. We tried a different scenario, where we first clean up all the legacy items, start the application with a big enough partiotion sizing and re-register all reminders during the app.
With this process, we see some GO-errors in the logs 
![image](https://user-images.githubusercontent.com/3714921/133208210-faeaaea3-ff7d-453b-adf1-468cface1230.png)
The underlying timer was ment to get executed at exactly `8:18am`, but no actions are taking place
